lib/ash/type/map.ex

```diff
@@ -31,11 +90,71 @@ defmodule Ash.Type.Map do
 
   @impl true
   def cast_stored(nil, _), do: {:ok, nil}
+
   def cast_stored(value, _) when is_map(value), do: {:ok, value}
+
   def cast_stored(_, _), do: :error
 
   @impl true
   def dump_to_native(nil, _), do: {:ok, nil}
   def dump_to_native(value, _) when is_map(value), do: {:ok, value}
   def dump_to_native(_, _), do: :error
+
+  @impl true
+  def apply_constraints(value, constraints) do
+    Enum.reduce(constraints, {:ok, value}, fn
+      {:fields, fields}, {:ok, value} ->
+        check_fields(value, fields)
+
+      {_, _}, {:error, errors} ->
+        {:error, errors}
+    end)
+  end
+
+  defp check_fields(value, fields) do
+    Enum.reduce(fields, {:ok, %{}}, fn
+      {field, field_constraints}, {:ok, checked_value} ->
+        case fetch_field(value, field) do
+          {:ok, nil} ->
+            if field_constraints[:allow_nil?] == false do
+              {:error, [[message: "value must not be nil", field: field]]}
+            else
+              {:ok, Map.put(checked_value, field, nil)}
+            end
+
+          {:ok, field_value} ->
+            case Ash.Type.apply_constraints(
```

Comment:
First we want to cast it.

```elixir
with {:ok, value} <- cast(...),
       {:ok, value} <- apply_constraints() do
end
```

Lastly, we need to move the `allow_nil?` check from above into this branch. A type might choose to cast `nil` -> `""` for example, or like `:string`, it might do `"" -> nil` meaning we need to do nil checks after casting.